From 225aa1df1738d4fb2b47e5fe7f566e63f82ec1ac Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Thu, 6 Mar 2025 15:53:28 +0100
Subject: [PATCH] drm/panel: himax-hx8394: add retries when dsi write failed

Some DSI writes do not always execute correctly and must be re-executed
to succeed. We will allow a few attempts.

Change-Id: I50dba151142ec21f5612b1d09d43d7b3775b531f
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/442561
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/gpu/drm/panel/panel-himax-hx8394.c | 121 ++++++++++++---------
 1 file changed, 72 insertions(+), 49 deletions(-)

--- a/drivers/gpu/drm/panel/panel-himax-hx8394.c
+++ b/drivers/gpu/drm/panel/panel-himax-hx8394.c
@@ -218,79 +218,102 @@ static const struct hx8394_panel_desc hs
 	.init_sequence = hsd060bhw4_init_sequence,
 };
 
+/*
+ * Some DSI writes do not always execute correctly and
+ * must be re-executed to succeed.
+ * We will allow a few attempts.
+ */
+#define MAX_DSI_WRITE_TRIES	5
+
+#define rocktech_rk055mhd042a0_dsi_write(dsi, cmd, seq...)                      \
+	do {                                                                    \
+		static const u8 d[] = { cmd, seq };                             \
+		int ret;                                                        \
+		int try;                                                        \
+		for (try = 0; try < MAX_DSI_WRITE_TRIES; try++) {               \
+			ret = mipi_dsi_dcs_write_buffer(dsi, d, ARRAY_SIZE(d)); \
+			if (ret >= 0)                                           \
+				break;                                          \
+		}                                                               \
+		if (ret < 0)                                                    \
+			return ret;                                             \
+	} while (0)
+
 static int rocktech_rk055mhd042a0_init_sequence(struct hx8394 *ctx)
 {
 	struct mipi_dsi_device *dsi = to_mipi_dsi_device(ctx->dev);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETEXTC,
-			       0xFF, 0x83, 0x94);
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETMIPI,
-			       0x61, 0x03, 0x68, 0x6B, 0xB2, 0xC0);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETEXTC, 0xFF, 0x83, 0x94);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETMIPI, 0x61, 0x03, 0x68, 0x6B,
+					 0xB2, 0xC0);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETPOWER,
-			       0x48, 0x12, 0x72, 0x09, 0x32, 0x54, 0x71, 0x71, 0x57, 0x47);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETPOWER, 0x48, 0x12, 0x72, 0x09,
+					 0x32, 0x54, 0x71, 0x71, 0x57, 0x47);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETDISP,
-			       0x00, 0x80, 0x64, 0x0C, 0x0D, 0x2F);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETDISP, 0x00, 0x80, 0x64, 0x0C,
+					 0x0D, 0x2F);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETCYC,
-			       0x73, 0x74, 0x73, 0x74, 0x73, 0x74, 0x01, 0x0C, 0x86, 0x75, 0x00,
-			       0x3F, 0x73, 0x74, 0x73, 0x74, 0x73, 0x74, 0x01, 0x0C, 0x86);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETCYC, 0x73, 0x74, 0x73, 0x74,
+					 0x73, 0x74, 0x01, 0x0C, 0x86, 0x75, 0x00, 0x3F,
+					 0x73, 0x74, 0x73, 0x74, 0x73, 0x74, 0x01, 0x0C,
+					 0x86);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETGIP0,
-			       0x00, 0x00, 0x07, 0x07, 0x40, 0x07, 0x0C, 0x00, 0x08, 0x10, 0x08,
-			       0x00, 0x08, 0x54, 0x15, 0x0A, 0x05, 0x0A, 0x02, 0x15, 0x06, 0x05,
-			       0x06, 0x47, 0x44, 0x0A, 0x0A, 0x4B, 0x10, 0x07, 0x07, 0x0C, 0x40);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETGIP0, 0x00, 0x00, 0x07, 0x07,
+					 0x40, 0x07, 0x0C, 0x00, 0x08, 0x10, 0x08, 0x00,
+					 0x08, 0x54, 0x15, 0x0A, 0x05, 0x0A, 0x02, 0x15,
+					 0x06, 0x05, 0x06, 0x47, 0x44, 0x0A, 0x0A, 0x4B,
+					 0x10, 0x07, 0x07, 0x0C, 0x40);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETGIP1,
-			       0x1C, 0x1C, 0x1D, 0x1D, 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06,
-			       0x07, 0x08, 0x09, 0x0A, 0x0B, 0x24, 0x25, 0x18, 0x18, 0x26, 0x27,
-			       0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18,
-			       0x18, 0x18, 0x18, 0x18, 0x18, 0x20, 0x21, 0x18, 0x18, 0x18, 0x18);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETGIP1, 0x1C, 0x1C, 0x1D, 0x1D,
+					 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+					 0x08, 0x09, 0x0A, 0x0B, 0x24, 0x25, 0x18, 0x18,
+					 0x26, 0x27, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18,
+					 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18,
+					 0x18, 0x18, 0x20, 0x21, 0x18, 0x18, 0x18, 0x18);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETGIP2,
-			       0x1C, 0x1C, 0x1D, 0x1D, 0x07, 0x06, 0x05, 0x04, 0x03, 0x02, 0x01,
-			       0x00, 0x0B, 0x0A, 0x09, 0x08, 0x21, 0x20, 0x18, 0x18, 0x27, 0x26,
-			       0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18,
-			       0x18, 0x18, 0x18, 0x18, 0x18, 0x25, 0x24, 0x18, 0x18, 0x18, 0x18);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETGIP2, 0x1C, 0x1C, 0x1D, 0x1D,
+					 0x07, 0x06, 0x05, 0x04, 0x03, 0x02, 0x01, 0x00,
+					 0x0B, 0x0A, 0x09, 0x08, 0x21, 0x20, 0x18, 0x18,
+					 0x27, 0x26, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18,
+					 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18,
+					 0x18, 0x18, 0x25, 0x24, 0x18, 0x18, 0x18, 0x18);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETVCOM,
-			       0x92, 0x92);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETVCOM, 0x92, 0x92);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETGAMMA,
-			       0x00, 0x0A, 0x15, 0x1B, 0x1E, 0x21, 0x24, 0x22, 0x47, 0x56, 0x65,
-			       0x66, 0x6E, 0x82, 0x88, 0x8B, 0x9A, 0x9D, 0x98, 0xA8, 0xB9, 0x5D,
-			       0x5C, 0x61, 0x66, 0x6A, 0x6F, 0x7F, 0x7F, 0x00, 0x0A, 0x15, 0x1B,
-			       0x1E, 0x21, 0x24, 0x22, 0x47, 0x56, 0x65, 0x65, 0x6E, 0x81, 0x87,
-			       0x8B, 0x98, 0x9D, 0x99, 0xA8, 0xBA, 0x5D, 0x5D, 0x62, 0x67, 0x6B,
-			       0x72, 0x7F,  0x7F);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETGAMMA, 0x00, 0x0A, 0x15, 0x1B,
+					 0x1E, 0x21, 0x24, 0x22, 0x47, 0x56, 0x65, 0x66,
+					 0x6E, 0x82, 0x88, 0x8B, 0x9A, 0x9D, 0x98, 0xA8,
+					 0xB9, 0x5D, 0x5C, 0x61, 0x66, 0x6A, 0x6F, 0x7F,
+					 0x7F, 0x00, 0x0A, 0x15, 0x1B, 0x1E, 0x21, 0x24,
+					 0x22, 0x47, 0x56, 0x65, 0x65, 0x6E, 0x81, 0x87,
+					 0x8B, 0x98, 0x9D, 0x99, 0xA8, 0xBA, 0x5D, 0x5D,
+					 0x62, 0x67, 0x6B, 0x72, 0x7F,  0x7F);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_UNKNOWN1, 0x1F, 0x31);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_UNKNOWN1, 0x1F, 0x31);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETPANEL, 0x03);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETPANEL, 0x03);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_UNKNOWN3, 0x02);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_UNKNOWN3, 0x02);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETREGBANK, 0x02);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETREGBANK, 0x02);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_UNKNOWN4,
-			       0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
-			       0xFF);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_UNKNOWN4, 0xFF, 0xFF, 0xFF, 0xFF,
+					 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETREGBANK, 0x00);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETREGBANK, 0x00);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETREGBANK, 0x01);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETREGBANK, 0x01);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETPOWER, 0x00);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETPOWER, 0x00);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_SETREGBANK, 0x00);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_SETREGBANK, 0x00);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_UNKNOWN0,
-			       0x40, 0x81, 0x50, 0x00, 0x1A, 0xFC, 0x01);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_UNKNOWN0, 0x40, 0x81, 0x50, 0x00,
+					 0x1A, 0xFC, 0x01);
 
-	mipi_dsi_dcs_write_seq(dsi, HX8394_CMD_UNKNOWN2, 0xED);
+	rocktech_rk055mhd042a0_dsi_write(dsi, HX8394_CMD_UNKNOWN2, 0xED);
 
-	mipi_dsi_dcs_write_seq(dsi, MIPI_DCS_SET_ADDRESS_MODE, FH);
+	rocktech_rk055mhd042a0_dsi_write(dsi, MIPI_DCS_SET_ADDRESS_MODE, FH);
 
 	return 0;
 }
